[DELETE with DVs] Allow reading CDF from files with Deletion Vectors - Part 2/2

xupefei opened the pull request:
## Description

This PR is part of https://github.com/delta-io/delta/issues/1701.

This is a follow-up of https://github.com/delta-io/delta/pull/1680 to add support to allow reading CDC from files that have DV associated. In this PR we modify the CDC reader to construct in-line DVs diff'ed from two existing DVs, and modify the corresponding FileIndex to use the in-line DV.

## How was this patch tested?

Adding a CDC variant to an existing suite.

xupefei commented:
Due to https://github.com/delta-io/delta/pull/1680 not being merged, the current PR looks bigger than what it actually is. The real diff is https://github.com/delta-io/delta/pull/1704/commits/0d183a2695aba2784895bc1e3e86ba95818d9f5f.

xupefei commented:
@vkorukanti Could you review?

vkorukanti commented on core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala:
```diff
@@ -154,6 +157,16 @@ object CDCReader extends CDCReaderImpl
       actions: Seq[T],
       commitInfo: Option[CommitInfo])
+
+  case class FileVersion(version: Long, timestamp: Timestamp) {
```
One or two lines doc describing what this is used for. Same comment for below case class.

xupefei replied:
Done!

vkorukanti commented on core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala:
```diff
@@ -608,31 +617,196 @@ trait CDCReaderImpl extends DeltaLogging {
+    // If a path is only being removed, then scan it normally as removed rows
+    (removeFilesMap.keySet -- addFilesMap.keySet).foreach { removeKey =>
+      finalRemoveFiles
+        .getOrElseUpdate(new FileVersion(removeKey), ListBuffer())
```
`FileVersion` seems like the wrong name? It looks like it is referring the the delta log version?

vkorukanti commented on core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala:
```diff
@@ -608,31 +617,196 @@ trait CDCReaderImpl extends DeltaLogging {
+          spark,
+          new CdcAddFileIndex(
+            spark,
+            Seq(CDCDataSpec(
```
nit: nested method for cdcDataSpec(fileVersion, addFiles, commitInfo) and use it here and below.

vkorukanti commented on core/src/main/scala/org/apache/spark/sql/delta/commands/cdc/CDCReader.scala:
```diff
@@ -751,6 +925,88 @@ trait CDCReaderImpl extends DeltaLogging {
+    //   b) Rows masked by DV1 but not DV2 are re-added. May happen when restoring a table.
+    (removeDvOpt, addDvOpt) match {
+      case (None, None) =>
+        throw new Exception("Unreadable code path reached.")
```
nit: add few more details to the error message: ex. same file is added and removed within the same version

xupefei replied:
Done!

vkorukanti commented on core/src/main/scala/org/apache/spark/sql/delta/files/CdcAddFileIndex.scala:
```diff
@@ -39,7 +39,8 @@ class CdcAddFileIndex(
     path: Path,
-    snapshot: Snapshot
+    snapshot: Snapshot,
+    override val rowIndexFilters: Option[Map[String, RowIndexFilterType]] = None
```
docs for the new param. Also mention the details such as the path should be a URI encoded in order for the look up to work.

vkorukanti replied:
same for remove. Also do we want to add assert to check if there is one entry for each file?

xupefei replied:
Done adding docs.

xupefei replied:
It's not necessary that `rowIndexFilters` contains all files in `filesByVersion`, the former is a superset of the latter. IMO asserting here won't bring too much add value.

vkorukanti commented on core/src/test/scala/org/apache/spark/sql/delta/DeltaCDCSuite.scala:
```diff
@@ -964,3 +964,11 @@ class DeltaCDCScalaSuite extends DeltaCDCSuiteBase {
 
 }
+
+class DeltaCDCScalaWithDeletionVectorsSuite extends DeltaCDCScalaSuite
```
does this suite cover the streaming tests?

xupefei replied:
Nope... but I just added one!

vkorukanti reviewed (changes requested):
LGTM, few comments.

vkorukanti commented:
If I understand correctly, the [DELETE with DVs](https://github.com/delta-io/delta/issues/1591) doesn't need to do anything special for CDF? If yes, please remove the CDF work item from the plan.

xupefei commented:
> If I understand correctly, the [DELETE with DVs](https://github.com/delta-io/delta/issues/1591) doesn't need to do anything special for CDF? If yes, please remove the CDF work item from the plan.

Thanks! I addressed all your comments.
I don't have the edit permission for the [DELETE with DVs](https://github.com/delta-io/delta/issues/1591) page, could you do it instead?
